cli/src/commands/run.rs

```diff
@@ -55,16 +51,16 @@ pub fn cmd_run(ui: &mut Ui, command: &CommandHelper, args: &RunArgs) -> Result<(
     let workspace_command = command.workspace_helper(ui)?;
     let _resolved_commits =
         resolve_multiple_nonempty_revsets(&args.revisions, &workspace_command, ui)?;
-    let _jobs = if let Some(_jobs) = args.jobs {
-        _jobs
-    } else {
-        // Use all available cores
-
-        // SAFETY:
-        // We use a internal constant of 4 threads, if it fails
-        let available =
-            std::thread::available_parallelism().unwrap_or(NonZeroUsize::new(4).unwrap());
-        available.into()
-    };
+    let _jobs = args
+        .jobs
+        .filter(|j| *j > 0)
+        .or_else(|| {
+            // Use all available cores
+            std::thread::available_parallelism()
+                .ok()
+                .and_then(|t| t.try_into().ok())
+        })
+        // Fallback to a single user-visible job.
+        .unwrap_or(1usize);
```

Comment:
```suggestion
    let jobs = match args.jobs {
        Some(0) | None => std::thread::available_parallelism().ok().and_then(|| t.try_into().ok()),
        Some(jobs) => Some(jobs),
    };
    let _jobs = jobs.unwrap_or(1usize);
```

- Personally, I think the use of `filter` here is hard to follow.
- After rewriting the code as in my suggestion, I realize that I think it doesn't do what we want: if the user doesn't pass `--jobs`, we should default to `1`, not the max parallelism. That is, we should only match on `Some(0)` for the max-parallelism case. (We don't want to e.g. spawn extra workspaces without the user explicitly opting-in, since that could use a lot of resources.)

Reply:
> * Personally, I think the use of `filter` here is hard to follow.

I'll add a comment to hopefully increase readability. 

> * After rewriting the code as in my suggestion, I realize that I think it doesn't do what we want: if the user doesn't pass `--jobs`, we should default to `1`, not the max parallelism. That is, we should only match on `Some(0)` for the max-parallelism case. (We don't want to e.g. spawn extra workspaces without the user explicitly opting-in, since that could use a lot of resources.)

I still disagree with only spawning a single "job", if we pass nothing for `-j`. I'm going to drop the argument now, but I'd rather spawn a "job" for each core then just default to one and if all else fails we use a single job. Your argument implies dropping `std::thread::available_parallelism()` completly. (I put job in quotes, as it probably is a process in user-terms and still implies something thread-like to me.)

The one thing I realised is that I want actual parallelism for the MVP of`run`, not concurrency. At some point in the future we then can rewrite it in terms of concurrency.

Reply:
I've added a comment for the `filter` part.